transport/src/main/java/io/netty/channel/ChannelInitializer.java

```diff
@@ -76,13 +89,48 @@ public final void channelRegistered(ChannelHandlerContext ctx) throws Exception
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
         logger.warn("Failed to initialize a channel. Closing: " + ctx.channel(), cause);
+        ctx.close();
+    }
+
+    /**
+     * {@inheritDoc} If override this method ensure you call super!
+     */
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        if (ctx.channel().isRegistered()) {
+            // This should always be true with our current DefaultChannelPipeline implementation.
+            // The good thing about calling initChannel(...) in handlerAdded(...) is that there will be no ordering
+            // suprises if a ChannelInitializer will add another ChannelInitializer. This is as all handlers
+            // will be added in the expected order.
+            initChannel(ctx);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private boolean initChannel(ChannelHandlerContext ctx) throws Exception {
+        if (initMap.putIfAbsent(ctx, Boolean.TRUE) == null) { // Guard against re-entrance.
```

Comment:
Can you clarify what the case is for re-entrance here? Don't both the add and registered events occur on the event loop thread?  Will one event trigger the other?